File: src/types/aggregate.ts
/**
 * Aggregated Result Type Definitions
 *
 * Aggregated results summarise multiple evaluation runs with statistical
 * measures. This is the intermediate format between raw results and
 * final rendered output (LaTeX tables, etc.).
 */

import type { Primitive } from "./case.js";
import type { SutRole } from "./sut.js";

/**
 * Summary statistics for a numeric metric.
 */
export interface SummaryStats {
	/** Number of observations */
	n: number;

	/** Arithmetic mean */
	mean: number;

	/** Median (50th percentile) */
	median: number;

	/** Minimum value */
	min: number;

	/** Maximum value */
	max: number;

	/** Standard deviation (sample) */
	std?: number;

	/** 95% confidence interval [lower, upper] */
	confidence95?: [number, number];

	/** Sum of all values */
	sum?: number;

	/** 25th percentile */
	p25?: number;

	/** 75th percentile */
	p75?: number;
}

/**
 * Comparison metrics between primary and baseline SUTs.
 */
export interface ComparisonMetrics {
	/** Absolute deltas (primary - baseline) */
	deltas: Record<string, number>;

	/** Ratios (primary / baseline) */
	ratios: Record<string, number>;

	/** Win rate (% of cases where primary beats baseline) */
	betterRate?: number;

	/** Mann-Whitney U statistic */
	uStatistic?: number;

	/** Statistical significance (p-value) */
	pValue?: number;

	/** Effect size (Cohen's d) */
	effectSize?: number;
}

/**
 * Coverage information for the aggregation.
 */
export interface CoverageMetrics {
	/** Fraction of cases covered */
	caseCoverage: number;

	/** Metric availability (metric name -> coverage fraction) */
	metricCoverage: Record<string, number>;

	/** Missing case IDs */
	missingCases?: string[];
}

/**
 * Aggregated result for a SUT (optionally grouped by case class).
 */
export interface AggregatedResult {
	/** SUT identifier */
	sut: string;

	/** SUT role */
	sutRole: SutRole;

	/** Case class (if grouped) */
	caseClass?: string;

	/** Grouping information */
	group: {
		/** Number of runs in this aggregate */
		runCount: number;

		/** Number of unique cases */
		caseCount: number;

		/** Hash of configuration (for homogeneity check) */
		configHash?: string;
	};

	/** Correctness summary */
	correctness: {
		/** Fraction of runs that produced valid output */
		validRate: number;

		/** Fraction of runs that produced any output */
		producedOutputRate: number;

		/** Fraction of runs matching expected (if oracle available) */
		matchesExpectedRate?: number;

		/** Breakdown of failure types */
		failureBreakdown?: Record<string, number>;
	};

	/** Aggregated metrics (metric name -> summary stats) */
	metrics: Record<string, SummaryStats>;

	/** Comparisons with baselines (baseline SUT id -> comparison) */
	comparisons?: Record<string, ComparisonMetrics>;

	/** Coverage information */
	coverage?: CoverageMetrics;

	/** Additional metadata */
	metadata?: Record<string, Primitive>;
}

/**
 * Complete aggregation output.
 */
export interface AggregationOutput {
	/** Schema version */
	version: string;

	/** Generation timestamp */
	timestamp: string;

	/** Aggregated results */
	aggregates: AggregatedResult[];

	/** Global metadata */
	metadata?: {
		/** Total runs processed */
		totalRuns: number;

		/** Total unique cases */
		totalCases: number;

		/** SUTs included */
		sutsIncluded: string[];

		/** Case classes included */
		caseClassesIncluded?: string[];
	};
}

File: python/src/ppef/types/aggregate.py
"""Aggregated result type definitions."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from .case import Primitive
from .sut import SutRole


class SummaryStats(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    n: int
    mean: float
    median: float
    min: float
    max: float
    std: float | None = None
    confidence95: tuple[float, float] | None = None
    sum: float | None = None
    p25: float | None = None
    p75: float | None = None


class ComparisonMetrics(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    deltas: dict[str, float]
    ratios: dict[str, float]
    better_rate: float | None = Field(default=None, alias="betterRate")
    u_statistic: float | None = Field(default=None, alias="uStatistic")
    p_value: float | None = Field(default=None, alias="pValue")
    effect_size: float | None = Field(default=None, alias="effectSize")


class CoverageMetrics(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    case_coverage: float = Field(alias="caseCoverage")
    metric_coverage: dict[str, float] = Field(alias="metricCoverage")
    missing_cases: list[str] | None = Field(default=None, alias="missingCases")


class AggregatedResult(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    sut: str
    sut_role: SutRole = Field(alias="sutRole")
    case_class: str | None = Field(default=None, alias="caseClass")
    group: dict[str, object]
    correctness: dict[str, object]
    metrics: dict[str, SummaryStats]
    comparisons: dict[str, ComparisonMetrics] | None = None
    coverage: CoverageMetrics | None = None
    metadata: dict[str, Primitive] | None = None


class AggregationOutput(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    version: str
    timestamp: str
    aggregates: list[AggregatedResult]
    metadata: dict[str, object] | None = None
